src/gui/src/drcWidget.cpp

```diff
@@ -0,0 +1,538 @@
+      idx_item->setEditable(false);
+      idx_item->setSelectable(false);
+
+      type_group->appendRow({
+        idx_item,
+        violation_item
+      });
+    }
+
+    QStandardItem* count_item = new QStandardItem(QString::number(violation_list.size()));
+    count_item->setEditable(false);
+    count_item->setSelectable(false);
+    model_->appendRow({type_group, count_item});
+  }
+
+  toggleRenderer();
+}
+
+void DRCWidget::drawObjects(Painter& painter)
+{
+  int min_box = 20.0 / painter.getPixelsPerDBU();
+  Painter::Color pen_color = Painter::white;
+  Painter::Color brush_color = pen_color;
+  brush_color.a = 50;
+
+  painter.setPen(pen_color, true, 0);
+  painter.setHashedBrush(brush_color);
+  for (const auto& violation : violations_) {
+    const odb::Rect& box = violation->getBBox();
+    if (std::max(box.dx(), box.dy()) < min_box) {
+      // box is too small to be useful, so draw X instead
+      odb::Point center(box.xMin() + box.dx() / 2, box.yMin() + box.dy() / 2);
+      painter.drawLine(
+          {center.x() - min_box / 2, center.y() - min_box / 2},
+          {center.x() + min_box / 2, center.y() + min_box / 2});
+      painter.drawLine(
+          {center.x() - min_box / 2, center.y() + min_box / 2},
+          {center.x() + min_box / 2, center.y() - min_box / 2});
+    } else {
+      violation->paint(painter);
+    }
+  }
+}
+
+SelectionSet DRCWidget::select(odb::dbTechLayer* layer, const odb::Point& point)
+{
+  if (layer != nullptr) {
+    return SelectionSet();
+  }
+
+  auto gui = Gui::get();
+
+  SelectionSet selections;
+  for (const auto& violation : violations_) {
+    if (violation->getBBox().intersects(point)) {
+      selections.insert(gui->makeSelected(violation.get()));
+    }
+  }
+  return selections;
+}
+
+void DRCWidget::loadReport(const QString& filename)
+{
+  violations_.clear();
+
+  if (filename.endsWith(".rpt")) {
+    loadTRReport(filename);
+  } else if (filename.endsWith(".ascii")) {
+    loadASCIIReport(filename);
+  } else {
+    logger_->error(utl::GUI, 32, "Unable to determine type of {}", filename.toStdString());
+  }
+
+  updateModel();
+}
+
+void DRCWidget::loadTRReport(const QString& filename)
+{
+  std::ifstream report(filename.toStdString());
+  if (!report.is_open()) {
+    logger_->error(utl::GUI, 30, "Unable to open TritonRoute DRC report: {}", filename.toStdString());
+  }
+
+  std::regex violation_type("\\s*violation type: (.*)");
+  std::regex srcs("\\s*srcs: (.*)");
+  std::regex bbox_layer("\\s*bbox = (.*) on Layer (.*)");
+  std::regex bbox_corners("\\s*\\(\\s*(.*),\\s*(.*)\\s*\\)\\s*-\\s*\\(\\s*(.*),\\s*(.*)\\s*\\)");
+
+  auto tech = block_->getDataBase()->getTech();
+  while (!report.eof()) {
+    std::string line;
+    std::smatch base_match;
+
+    // type of violation
+    std::getline(report, line);
+    std::string type;
+    if (std::regex_match(line, base_match, violation_type)) {
+      type = base_match[1].str();
+    } else {
+      continue;
```

Comment:
The continue is really an error in parsing and warrants a warning.  Here and below.